File: src/main/java/com/pseuco/cp24/slug/Slug.java
package com.pseuco.cp24.slug;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

import com.pseuco.cp24.Config;
import com.pseuco.cp24.request.CustomerId;
import com.pseuco.cp24.request.Request;
import com.pseuco.cp24.request.RequestHandler;
import com.pseuco.cp24.request.ServerId;

/**
 * A slow request handler processing requests sequentially.
 */
public class Slug implements RequestHandler {
    /**
     * Server ID of this single server implementation.
     */
    private final ServerId serverId = ServerId.generate();

    /**
     * Stack of tickets still available.
     */
    private final Stack<Ticket> available;

    /**
     * Timeout of reservations in seconds.
     */
    private final int timeout;

    /**
     * Reservations made by customers.
     */
    private Map<CustomerId, Reservation> reservations = new HashMap<>();

    /**
     * Constructs a new slug.
     *
     * @param config Options for the sales system.
     */
    public Slug(final Config config) {
        this.available = Ticket.generateStack(config.getNumTickets());
        this.timeout = config.getTimeout();
    }

    /**
     * Aborts and removes reservations with an expired timeout.
     */
    private void clearReservations() {
        this.reservations.values().removeIf(reservation -> {
            if (reservation.getAge() > this.timeout) {
                // Make the ticket available again.
                this.available.push(reservation.abort());
                return true;
            } else {
                return false;
            }
        });
    }

    @Override
    public synchronized void handle(final Request request) {
        // We first need to clear reservations with an expired timeout. In your
        // implementation, you need to do something equivalently locally on each
        // server.
        this.clearReservations();

        // Now, we can handle the request.
        switch (request.getKind()) {
            // This request is handled by the load balancer.
            case NUM_SERVERS -> {
                switch (request.getMethod()) {
                    case GET -> {
                        // In your implementation, you need to respond with the number of servers.
                        request.respondWithInt(1);
                    }
                    case POST -> {
                        final var numServers = request.readInt();
                        if (numServers.isEmpty()) {
                            // The client is supposed to provide a number of servers.
                            request.respondWithError("No number of servers provided!");
                        } else {
                            // In your implementation, you need to support this request for on-demand
                            // scaling. After
                            // scaling, you should respond with the number of servers.
                            request.respondWithError("Slug does not support on-demand scaling!");
                        }
                    }
                }
            }
            // This request is handled by the load balancer.
            case GET_SERVERS -> {
                // In your implementation, you need to respond with the IDs of the active
                // servers.
                final ServerId[] ids = { this.serverId };
                request.respondWithServerIds(Arrays.asList(ids));
            }

            // In your implementation, this request need to be redirected to a server.
            case NUM_AVAILABLE_TICKETS -> {
                // This request requires us to respond with a server ID.
                request.setServerId(this.serverId);
                // In your implementation, you respond with an approximation of the actual
                // number.
                request.respondWithInt(this.available.size());
            }

            // In your implementation, this request need to be redirected to a server.
            case RESERVE_TICKET -> {
                // This request requires us to respond with a server ID.
                request.setServerId(this.serverId);

                final var customer = request.getCustomerId();
                if (this.reservations.containsKey(customer)) {
                    // We do not allow a customer to reserve more than a ticket at a time.
                    request.respondWithError("A ticket has already been reserved!");
                } else if (!this.available.isEmpty()) {
                    // Take a ticket from the stack of available tickets and reserve it.
                    final var ticket = this.available.pop();
                    this.reservations.put(customer, new Reservation(ticket));
                    // Respond with the ID of the reserved ticket.
                    request.respondWithInt(ticket.getId());
                } else {
                    // Tell the client that no tickets are available.
                    request.respondWithSoldOut();
                }
            }
            // In your implementation, this request need to be redirected to a server.
            case ABORT_PURCHASE -> {
                // This request requires us to respond with a server ID.
                request.setServerId(this.serverId);

                final var customer = request.getCustomerId();
                if (!this.reservations.containsKey(customer)) {
                    // Without a reservation there is nothing to abort.
                    request.respondWithError("No ticket has been reserved!");
                } else {
                    final var reservation = this.reservations.get(customer);
                    final var ticketId = request.readInt();
                    if (ticketId.isEmpty()) {
                        // The client is supposed to provide a ticket ID.
                        request.respondWithError("No ticket ID provided!");
                    } else if (ticketId.get() == reservation.getTicketId()) {
                        // Abort the reservation and put the ticket back on the stack.
                        final var ticket = reservation.abort();
                        this.available.push(ticket);
                        this.reservations.remove(customer);
                        // Respond with the ID of the formerly reserved ticket.
                        request.respondWithInt(ticket.getId());
                    } else {
                        // ID does not match the ID of the reservation.
                        request.respondWithError("Invalid ticket ID provided!");
                    }
                }
            }
            // In your implementation, this request need to be redirected to a server.
            case BUY_TICKET -> {
                // This request requires us to respond with a server ID.
                request.setServerId(this.serverId);

                final var customer = request.getCustomerId();
                if (!this.reservations.containsKey(customer)) {
                    // Without a reservation there is nothing to buy.
                    request.respondWithError("No ticket has been reserved!");
                } else {
                    final var reservation = this.reservations.get(customer);
                    final var ticketId = request.readInt();
                    if (ticketId.isEmpty()) {
                        // The client is supposed to provide a ticket ID.
                        request.respondWithError("No ticket ID provided!");
                    } else if (ticketId.get() == reservation.getTicketId()) {
                        // Sell the ticket to the customer.
                        final var ticket = reservation.sell();
                        this.reservations.remove(customer);
                        // Respond with the ID of the sold ticket.
                        request.respondWithInt(ticket.getId());
                    } else {
                        // ID does not match the ID of the reservation.
                        request.respondWithError("Invalid ticket ID provided!");
                    }
                }
            }

            // Use this request for sending debug information of your choosing.
            case DEBUG -> {
                request.respondWithString("This is 🐌.");
            }
        }
    }

    @Override
    public void shutdown() {
        // nothing to do
    }
}

File: crates/ticket-sale-server/src/slug.rs
//! 🐌 A sequential implementation for reference

use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::time::Instant;

use parking_lot::Mutex;
use ticket_sale_core::Config;
use ticket_sale_core::{Request, RequestHandler, RequestKind};
use uuid::Uuid;

#[derive(Debug)]
struct Reservation {
    ticket: u32,
    reserved_at: Instant,
}

impl Reservation {
    #[inline]
    fn new(ticket: u32) -> Self {
        Self {
            ticket,
            reserved_at: Instant::now(),
        }
    }

    #[inline]
    fn age_secs(&self) -> u64 {
        self.reserved_at.elapsed().as_secs()
    }
}

struct ServerInner {
    /// The server's ID
    id: Uuid,

    /// List of available ticket IDs
    available_tickets: Vec<u32>,

    /// Reservations made by customers
    reservations: HashMap<Uuid, Reservation>,

    /// Reservation timeout in seconds
    reservation_timeout: u32,
}

impl ServerInner {
    /// Abort and remove expired reservations
    fn clear_reservations(&mut self) {
        self.reservations.retain(|_, res| {
            if res.age_secs() > self.reservation_timeout as u64 {
                self.available_tickets.push(res.ticket);
                false
            } else {
                true
            }
        });
    }
}

impl ServerInner {
    fn handle(&mut self, mut rq: Request) {
        self.clear_reservations();

        match rq.kind() {
            // In your implementation, the following requests will be handled by the coordinator.
            RequestKind::GetNumServers => {
                // In your implementation, you need to respond with the number
                // of servers.
                rq.respond_with_int(1);
            }
            RequestKind::SetNumServers => {
                if rq.read_u32().is_some() {
                    // In your implementation, you need to support this request
                    // for on-demand scaling. After scaling, you should respond
                    // with the number of servers.
                    rq.respond_with_err("Slug does not support on-demand scaling!");
                } else {
                    rq.respond_with_err("No number of servers provided!");
                }
            }
            RequestKind::GetServers => rq.respond_with_server_list(&[self.id]),

            // Handling the following requests will remain the Server's responsibility.
            RequestKind::NumAvailableTickets => {
                // This request requires us to respond with a server id
                rq.set_server_id(self.id);

                rq.respond_with_int(self.available_tickets.len() as u32)
            }
            RequestKind::ReserveTicket => {
                // This request requires us to respond with a server id
                rq.set_server_id(self.id);

                let customer = rq.customer_id();
                match self.reservations.entry(customer) {
                    Entry::Occupied(_) => {
                        // We do not allow a customer to reserve more than a
                        // ticket at a time.
                        rq.respond_with_err("A ticket has already been reserved!");
                    }
                    Entry::Vacant(entry) => {
                        // Try to take a ticket from the stack of available
                        // tickets and reserve it.
                        if let Some(ticket) = self.available_tickets.pop() {
                            entry.insert(Reservation::new(ticket));
                            // Respond with the id of the reserved ticket
                            rq.respond_with_int(ticket);
                        } else {
                            // Tell the client that no tickets are available.
                            rq.respond_with_sold_out();
                        }
                    }
                }
            }
            RequestKind::BuyTicket => {
                // This request requires us to respond with a server id
                rq.set_server_id(self.id);

                if let Some(ticket) = rq.read_u32() {
                    let cid = rq.customer_id();
                    if let Some(res) = self.reservations.get(&cid) {
                        if ticket != res.ticket {
                            // The id does not match the id of the reservation
                            rq.respond_with_err("Invalid ticket id provided!");
                        } else {
                            // Sell the ticket to the customer
                            self.reservations.remove(&cid);
                            rq.respond_with_int(ticket);
                        }
                    } else {
                        // Without a reservation there is nothing to buy.
                        rq.respond_with_err("No ticket has been reserved!")
                    }
                } else {
                    // The client is supposed to provide a ticket id.
                    rq.respond_with_err("No ticket id provided!");
                }
            }
            RequestKind::AbortPurchase => {
                // This request requires us to respond with a server id
                rq.set_server_id(self.id);

                if let Some(ticket) = rq.read_u32() {
                    let cid = rq.customer_id();
                    if let Some(res) = self.reservations.get(&cid) {
                        if ticket != res.ticket {
                            // The id does not match the id of the reservation.
                            rq.respond_with_err("Invalid ticket id provided!");
                        } else {
                            // Abort the reservation and put the ticket back on the stack.
                            self.reservations.remove(&cid);

                            self.available_tickets.push(ticket);
                            // Respond with the id of the formerly reserved ticket.
                            rq.respond_with_int(ticket);
                        }
                    } else {
                        // Without a reservation there is nothing to abort.
                        rq.respond_with_err("No ticket has been reserved!")
                    }
                } else {
                    // The client is supposed to provide a ticket id.
                    rq.respond_with_err("No ticket id provided!");
                }
            }

            // Debug requests you may freely use to debug your implementation.
            RequestKind::Debug => {
                // 📌 Hint: You can use `rq.url()` and `rq.method()` to
                // implement multiple debugging commands.
                rq.respond_with_string("This is 🐌.");
            }
        }
    }
}

/// A request handler processing requests sequentially
pub struct Server(Mutex<ServerInner>);

impl RequestHandler for Server {
    fn handle(&self, request: Request) {
        self.0.lock().handle(request)
    }

    fn shutdown(self) {
        // nothing to do
    }
}

impl Server {
    /// Create a new slug
    pub fn new(config: &Config) -> Self {
        let inner = ServerInner {
            id: Uuid::new_v4(), // random uuid
            available_tickets: (0..config.tickets).collect(),
            reservations: HashMap::new(),
            reservation_timeout: config.timeout,
        };
        Self(Mutex::new(inner))
    }
}
